### DISCUSSION TITLE
How to make private LuaRocks repository server

### DISCUSSION THREAD

**Discussion by 0xbillw** [keplerproject/luarocks#1601] [Q&A]
I think Luarocks is the ideal package management tool in Lua.
But so far I've found that I can only upload modules to "[luarocks.org"](https://luarocks.org/).
Yes, there is _upload_servers_ parameter configurable for Luarocks, but I haven't found any open source Luarocks server programs available.
If you're doing your development in the Java programming language, a good package management tool is [Maven](http://maven.apache.org/).There are many open source components available for [Maven](http://maven.apache.org/), such as [Nexus](https://www.sonatype.com/nexus-repository-oss), [ArtiFactory](https://jfrog.com/open-source/#artifactory).

**Comment by Tieske** [comment]
the LuaRocks website is also on github, so you can run your own.

**Comment by franknarf8** [comment]
There is a ticket to support LuaRocks in Artifactory : https://www.jfrog.com/jira/browse/RTFACT-19711
But it doesn't seem to have a lot of traction so far...

Would it be possible to use a Generic repository type in Artifactory to store LoaRocks? Or does the package manager require specific API endpoint on the server?

**Comment by hishamhm** [comment]
You don't even need to run a clone of luarocks.org as Thijs suggested.

A rocks server just needs to be a HTTP (or `file://`!) accessible URL containing rock or rockspec files, and a manifest generated via `luarocks-admin make-manifest`. Nothing else special is needed. Other projects in the past [have used GitHub repos as private rocks servers](https://github.com/torch/rocks). I am aware of doing this as well.

Of course, `luarocks upload` won't work in those cases if it doesn't implement the upload API, but you can just upload files to your server by any other means, as long as you keep the manifest file up-to-date.

**Reply by sewbacca** [reply]
Is the upload protocol documented somewhere? I wonder how hard it would be to implement a simple serverside script for that.

**Reply by Tieske** [reply]
I did a manual upload here: https://github.com/Kong/lua-resty-aws/blob/main/upload.sh
